base_layer/wallet/src/ffi.rs

```diff
@@ -120,284 +435,574 @@ pub unsafe extern "C" fn create_pending_transaction_outputs(
+pub unsafe extern "C" fn wallet_get_balance(wallet: *mut Wallet) -> c_ulonglong {
+    //(*wallet).getBalance();
+    return 0;
+}
 
 // Create and send the first stage of a transaction to the specified wallet for the specified amount and with the
 // specified fee.
 #[no_mangle]
-pub unsafe extern "C" fn send_transaction(
-    wallet: *mut WalletFfi,
-    amount: c_ulonglong,       // MicroTari
-    fee_per_gram: c_ulonglong, // MicroTari
-    lock_height: c_ulonglong,  // u64
-    // Destination Node Peer
-    public_key: *const c_char,  // Byte[32] - PublicKey
-    net_address: *const c_char, // NetAddress
+pub unsafe extern "C" fn wallet_send_transaction(
+    wallet: *mut Wallet,
+    peer: *mut Peer,
+    transaction: *mut Transaction,
 ) -> bool
 {
-    // This function will need to check if the peer already exists and if not create it before sending
-    // A callback will be used by LibWallet to send the resultant data back to the client for storage.
+    if wallet.is_null() {
+        return false;
+    }
+
+    if transaction.is_null() {
+        return false;
+    }
+
+    if peer.is_null() {
+        return false;
+    }
+    //(*wallet).sendTransaction((*peer),(*transaction)) ? true : false
+    return true;
 }
 
 /// Cancel a pending outbound transaction so that the wallet will not complete and broadcast it if a reply is received
 #[no_mangle]
-pub unsafe extern "C" fn cancel_transaction(wallet: *mut WalletFfi, tx_id: c_ulonglong) -> bool {}
+pub unsafe extern "C" fn wallet_cancel_transaction(wallet: *mut Wallet, tr: *mut Transaction) -> bool {
+    if wallet.is_null() {
+        return false;
+    }
+
+    if tr.is_null() {
+        return false;
+    }
+
+    //(*wallet).cancelTransaction ((*tr)) ? true : false
+    return true;
+}
+
+// Callback Definition - Example
+
+// Will probably have to implement as a struct of callbacks in wallet, with wallet only calling the
+// functions if they are callable from the relevant wallet function, where the register callback functions
+// will bind the relevant c equivalent funciton pointer to the associated function
+// The Rust
+//
+// use std::os::raw::{c_int, c_uchar};
+//
+// #[no_mangle]
+// pub struct MyState {
+// pub call_back: extern "C" fn(*const c_uchar) -> c_int
+// }
+//
+// #[no_mangle]
+// pub extern fn get_state(call: extern "C" fn(*const c_uchar) -> c_int) -> *const () {
+// let state = MyState { call_back: call };
+// Box::into_raw(Box::new(state)) as *const _
+// }
+//
+// #[no_mangle]
+// pub extern fn run(state: *mut MyState) -> c_int {
+// unsafe {
+// ((*state).call_back)(format!("Callback run").as_ptr())
+// }
+// }
+//
+// #[no_mangle]
+// pub extern fn delete_state(state: *mut MyState) {
+// unsafe {
+// Box::from_raw(state);
+// }
+// }
+//
+// The C
+// #include <iostream>
+//
+// extern "C" {
+// void* get_state(int (*callback)(char*));
+// int run(void* state);
+// void delete_state(void* state);
+// }
+//
```

Comment:
Now that the data backends are abstracted I will shortly revisit the callback signatures below but in the meantime here is a question:

What does the Byte[32] argument look like in the C callback function signature i.e. the spending key in the following function header.
```C
// Append an output to be spent onto an existing PendingTransactionOutputs record
 int add_output_to_be_spent(	// int add_output_to_be_spent(
      ulonglong tx_id,	//      ulonglong tx_id,
      ulonglong value,	//      ulonglong value,
      *char spending_key,	//      *char spending_key,
      uchar feature_flags,	//      uchar feature_flags,
     ulonglong maturity	//      ulonglong maturity
 ) {}	
```

Reply:
Would be similar to the following:
For Rust
pub struct ByteVector(Vec<c_uchar>);
pub call_back: extern "C" fn(*mut ByteVector) // (1) without return from callback, i.e fire and forget
pub call_back: extern "C" fn(*mut ByteVector) -> c_int // (2) with return int back from callback
For C
void* ((void*) (*callback)(char*)); // (1)
void* (int (*callback)(char*)); // (2)

Then C side will use functions exposed from ByteVector pointer to get the values and length